[Flang] Move builtin .mod generation into runtimes (Reapply #137828)

Meinersbur opened the pull request:
Reapplication of #137828 and #169638

Changes on top of #169638:

  * [`9e7ab48`](https://github.com/llvm/llvm-project/pull/171515/commits/9e7ab487945a04468469e82aab81cbf86a2da0d7): When using an offloading toolchain (amdgpu, nvptx, ...), the driver also adds the host platform's builtin module directory to the search path. For instance:
```sh
flang -fintrinsic-modules-path ../lib/clang/22/finclude/flang/amdgcn-amd-amdhsa  -fintrinsic-modules-path  ../lib/clang/22/finclude/flang/x86_64-unknown-linux-gnu
```
So if `omp_lib.mod` is not available in `amdgcn-amd-amdhsa/`, it can use the one in `x86_64-unknown-linux-gnu/`. This means existing build setups don't need to be changed (`-DLLVM_RUNTIME_TARGETS=amdgcn-amd-amdhsa -DLLVM_ENABLE_RUNTIMES=openmp`). `omp_lib.f90` is not specific to any platform (only has `#if __INTEL_COMPILER`).

 * Split `RUNTIMES_FLANG_MODULES_ENABLED` option into `RUNTIMES_ENABLE_FORTRAN` and `RUNTIMES_FORTRAN_MODULES`.
 
   `RUNTIMES_ENABLE_FORTRAN` controls whether Fortran sources can be compiled to object (and module) files while `RUNTIMES_FORTRAN_MODULES` determines whether these can be used by Flang. There are two reasons for this change: 
   (1) Previously openmp had code to compile modules for gfortran and that possibility should be preserved. In that case modules files should not be installed into Flang's module path.
   (2) Make building the library (`*_ENABLE_STATIC`/`*_ENABLE_SHARED`) and the module (`*_FORTRAN_MODULES`) configurable separately. `RUNTIMES_FORTRAN_MODULES` is now the default value for  `LIBOMP_FORTRAN_MODULES`[^1] and `FLANG_RT_FORTRAN_MODULES`

[^1]: `LIBOMP_FORTRAN_MODULES` already exists before this PR and the other names are defined from that

 * After #197687 no object file output from compiling a Fortran file is needed for the libraries anymore. We now completely split the targets that build the libraries (flang_rt.runtime, etc) from the targets that build the modules (flang-rt-mod). Flang is not needed anymore to build the libraries, but only with `FLANG_RT_FORTRAN_MODULES=ON`. Modules continue to be required for `check-flang`. See #198205 to actually build flang-rt without building Flang first.


## General description

Move building the .mod files from openmp/flang to openmp/flang-rt using a shared mechanism. Motivations to do so are:

1. Most modules are target-dependent and need to be re-compiled for each target separately, which is something the LLVM_ENABLE_RUNTIMES system already does. Prime example is `iso_c_binding.mod` which encodes the target's ABI. Constants such as [`c_long_double` also have different values](https://github.com/llvm/llvm-project/blob/d748c81218bee39dafb9cc0c00ed7831a3ed44c3/flang-rt/lib/runtime/iso_c_binding.f90#L77-L81). Most other modules have `#ifdef`-enclosed code as well. For instance this caused offload targets nvptx64-nvidia-cuda/amdgpu-amd-amdhsa to use the modules files compiled for the host which may contrain uses of the types REAL(10) or REAL(16) not available for nvptx/amdgpu.

#146876
#128015
#129742
#158790

2. CMake has support for Fortran that we should use. Among other things, it automatically determines module dependencies so there is no need to hardcode them in the CMakeLists.txt.

Currently the dependencies between builtin modules and `omp_lib.mod` seem to be incorrect: https://github.com/llvm/llvm-project/pull/174474#issuecomment-3880189581

3. It allows using Fortran itself to implement Flang-RT. Currently, only `iso_fortran_env_impl.f90` emits object files that are needed by Fortran applications (#89403). The workaround of #95388 could be reverted (PR #169525).


If using Flang for cross-compilation or target-offloading, flang-rt must now be compiled for each target not only for the library, but also to get the target-specific module files. For instance in a bootstrapping runtime build, this can be done by adding: `-DLLVM_RUNTIME_TARGETS=default;nvptx64-nvidia-cuda;amdgpu-amd-amdhsa`.


Some new dependencies come into play:
 * openmp depends on flang-rt for building `lib_omp.mod` and `lib_omp_kinds.mod`. Currently, if flang-rt is not found then the modules are not built.
 * check-flang depends on flang-rt: If not found, the majority of tests are disabled. If not building in a bootstrpping build, the location of the module files can be pointed to using `-DFLANG_INTRINSIC_MODULES_DIR=<path>`, e.g. in a flang-standalone build. Alternatively, the test needing any of the intrinsic modules could be marked with `REQUIRES: flangrt-modules`.
 * check-flang depends on openmp: Not a change; tests requiring `lib_omp.mod` and `lib_omp_kinds.mod` those are already marked with `openmp_runtime`.

As intrinsic are now specific to the target, their location is moved from `include/flang` to `<resource-dir>/finclude/flang/<triple>`. The mechnism to compute the location have been moved from flang-rt (previously used to compute the location of `libflang_rt.*.a`) to common locations in `cmake/GetToolchainDirs.cmake` and `runtimes/CMakeLists.txt` so they can be used by both, openmp and flang-rt. Potentially the mechnism could also be shared by other libraries such as compiler-rt. 

`finclude` was chosen because `gfortran` uses it as well and avoids misuse such as `#include <flang/iso_c_binding.mod>`. The search location is now determined by `ToolChain` in the driver, instead of by the frontend. Another subdirectory `flang` avoid accidental inclusion of gfortran-modules which due to compression would result in user-unfriendly errors. Now the driver adds `-fintrinsic-module-path` for that location to the frontend call (Just like gfortran does). `-fintrinsic-module-path` had to be fixed for this because ironically it was only added to `searchDirectories`, but not `intrinsicModuleDirectories_`. Since the driver determines the location, tests invoking `flang -fc1` and `bbc` must also be passed the location by llvm-lit. This works like llvm-lit does for finding the include dirs for Clang using `-print-file-name=...`.

Meinersbur committed:
Reapply "[Flang] Move builtin .mod generation into runtimes (Reapply #137828)"

This reverts commit d233e787f0adfa2acd2e6c67aa2c362f9cf47ab4.

Meinersbur committed:
Move GetToolchainDirs.cmake

Meinersbur committed:
Backport runtimes files

Meinersbur committed:
Avoid redundancies in flang-rt CMakeLists.txt

Meinersbur committed:
Avoid whitespace-only change

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur commented:
@petrhosek I would appreciate if we can make progress in this PR. The original PR has already been approved by others.

Meinersbur committed:
[Driver] Fallback to host builtin modules

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/flang_builtin-mods_3' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Avoid general references to Flang

Meinersbur committed:
Re-add fixme removed from runtimes/CMakeLists.txt

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge commit '45f446837f00' into users/meinersbur/flang_builtin-mods_3

vzakhari commented:
Hi Michael, do you have an approximate ETA for this merge? I see you are making more changes, so I am just wondering if you think they will keep coming or you are close to the finish.

Meinersbur commented:
Not this year anymore, our operations would not be able to react before the holidays if something goes wrong. After that, it depends on @petrhosek's review. I don't know whether they will have larger change requests. It's been NFC only so far.

Changes made are responding to @petrhosek's review comments (also from #171610), with the exception of [`9e7ab48` (#171515)](https://github.com/llvm/llvm-project/pull/171515/commits/9e7ab487945a04468469e82aab81cbf86a2da0d7) which I made in the hope to reduce build configuration problems.

Meinersbur committed:
Remove unused include again

Meinersbur committed:
Address Petr Hosek's review

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

vzakhari commented:
Hi @Meinersbur, thanks for the ping!

Let me try the patches on my side. Do you plan to merge them in a particular order or/and merge them at the same time?

vzakhari commented:
Hi @Meinersbur, we are ready to land the two patches safely on our side. Thank you!

Meinersbur committed:
Merge branch 'main' into users/meinersbur/runtimes_flangrt-common

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Fix bug2021.f90

Meinersbur committed:
Fix dir name case

Meinersbur committed:
Try fixing libcxx-ci build issue

Meinersbur committed:
Build fix

Meinersbur committed:
Address review comments

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
rename config-runtimes-Fortran.cmake -> config-Fortran.cmake

Meinersbur committed:
Consecutive Fortran-handling cmake code

Meinersbur committed:
avoid unrelated change

Meinersbur committed:
Introduce RUNTIMES_*_PATH/DIR logic for all runtimes

Meinersbur committed:
extract common code

Meinersbur committed:
Add TODO comments

Meinersbur committed:
Add include for standalone build

Meinersbur committed:
typos

Meinersbur committed:
Try fix PR

Meinersbur committed:
Address review comments by @petrhosek

Meinersbur committed:
Merge branch 'main' into users/meinersbur/runtimes_resource-dir

Meinersbur committed:
Unbreak test

Meinersbur committed:
Merge branch 'main' into users/meinersbur/runtimes_resource-dir

Meinersbur committed:
Address some of @petrhosek's review remarks

tarunprabhu commented:
It looks like this PR will address some of the issues being faced by the authors of #182230. It would help if we could make some progress on this.

s-watanabe314 commented:
The problem I am currently facing in #182230 is precisely due to the .mod files not being created per target. This PR should be able to solve that, so merging it would be a great help.

Meinersbur committed:
Update file documentation

Meinersbur committed:
Merge commit 'f8906704104e446a7482aeca32d058b91867e05c' into users/meinersbur/runtimes_flangrt-common

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/runtimes_resource-dir' into users/meinersbur/runtimes_flangrt-common

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Reapply c4170461d7fc5ddd0423c5f7325abd0a7dfc6277

Meinersbur committed:
Remove unused LIBOMP_FORTRAN_MODULES

Meinersbur committed:
Merge branch 'main' into HEAD

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_resource-dir' into HEAD

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into HEAD

Meinersbur committed:
Use -cpp

Meinersbur committed:
Test fix

Meinersbur committed:
Merge commit '63c9573f1c6bafbf4dc57c470cbef13842283b58' into HEAD

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Merge commit '840fa81c691449823422552c4bf65a2d336ac2d1' into HEAD

Meinersbur committed:
Apply changes from runtimes_resource-dir

Meinersbur committed:
Merge commit 'c8280ddb02dab6df5897bc965a5fd95f382d31b2' into HEAD

Meinersbur committed:
Avoid unnecessary changes

Apply changes from runtimes_resource-dir

Forward-port

cleanup

cleanup

Remove RUNTIMES_DEVICE_ARCHITECTURES

Meinersbur committed:
Merge commit 'e69c7312f31be57af5ef17ada7466b46f8519f2a' into HEAD

Meinersbur committed:
Merge commit '9b1be8e3a497119920d60614a834415e517a8c36' into HEAD

Meinersbur committed:
Merge commit '723872cf551cc75134d2522042eae7c32941482d' into HEAD

Meinersbur committed:
Cleanup

gfortran compatibility

Meinersbur committed:
Merge commit 'b164e7c61020aa40995779574bb7e8064f165a09' into HEAD

Meinersbur committed:
Merge commit '6854802a3dc90c8344e64bfa03dff6e839c8c59f' into HEAD

Meinersbur committed:
Merge commit 'cfeee082b671e2ac6ee6c24babddf5dc89dd264c' into HEAD

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Avoid change noise

Meinersbur committed:
More change noise avoidance

Meinersbur committed:
Merge commit 'f1b0a84003a03580ed4473e1f4201a415ddaf0fd' into HEAD

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Merge commit '100c020e455aaba88b24693d215b6f62d82c6ace' into HEAD

Meinersbur committed:
Merge commit 'cc9c715e88ed233d3c5ae2114822b985938f60b4' into HEAD

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Fix incomplete merge

Meinersbur committed:
Merge commit '1e19b4364dd3f827e4110b0bc14ec31bf5bbaf59' into HEAD

Meinersbur committed:
Merge commit '61a43720f3e31357ff3842a02d5460e71e4062a6' into HEAD

Meinersbur committed:
Backport TODO comment

Meinersbur committed:
Merge commit '4214705219b0632304ab25b1b1cf9d6ebff6a589' into HEAD

Meinersbur committed:
Reduce change noise accross stacked PRs

Backport TODO comment

Meinersbur committed:
Merge commit '7c4e03a06bcfd80c6946217032a7509d797ee9ea' into HEAD

Meinersbur committed:
Reduce change noise accross stacked PRs

Add feedback if libomp-mod is used even without Fortran support

Reduce change noise accross stacked PRs

Backport TODO comment

Reuse LIBOMP_FORTRAN_MODULES

Meinersbur committed:
Merge commit '803d1d66094c9d2e543f5e62a426ef972e100a93' into HEAD

Meinersbur committed:
Merge commit '7389131f6843fd6110f98e1d102387fb576ff2ef' into HEAD

Meinersbur committed:
Merge commit '2bec75c4d6d1bcd369c16055402959636e241336' into HEAD

Meinersbur committed:
Merge branch 'main' into users/meinersbur/runtimes_flangrt-common

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Split RUNTIMES_FLANG_MODULES_ENABLED into RUNTIMES_ENABLE_FORTRAN/RUNTIMES_ENABLE_FLANG_MODULES

Meinersbur committed:
Split RUNTIMES_FLANG_MODULES_ENABLED into RUNTIMES_ENABLE_FORTRAN/RUNTIMES_ENABLE_FLANG_MODULES

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Add missing get_toolchain_module_subdir

Meinersbur committed:
Merge branch 'main' into users/meinersbur/runtimes_flangrt-common

Meinersbur committed:
Partial inlining of config-Fortran

Meinersbur committed:
Assume no Fortran in other runtimes

Ensure non-empty dependency list

Meinersbur committed:
Merge branch 'users/meinersbur/runtimes_flangrt-common' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Can assume non-empty RUNTIMES_FORTRAN_BUILD_DEPS

arsenm commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6755,6 +6755,17 @@ std::string Driver::GetFilePath(StringRef Name, const ToolChain &TC) const {
   if (llvm::sys::fs::exists(Twine(P)))
     return std::string(P);
 
+  // With Flang, also look for instrinsic modules
```
```suggestion
  // With Flang, also look for intrinsic modules
```

arsenm commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6755,6 +6755,17 @@ std::string Driver::GetFilePath(StringRef Name, const ToolChain &TC) const {
+            TC.getDefaultIntrinsicModuleDir()) {
+      SmallString<128> P(*IntrPath);
+      llvm::sys::path::append(P, Name);
+      if (llvm::sys::fs::exists(Twine(P)))
```
```suggestion
      if (llvm::sys::fs::exists(P))
```

arsenm commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -6755,6 +6755,17 @@ std::string Driver::GetFilePath(StringRef Name, const ToolChain &TC) const {
+  if (IsFlangMode()) {
+    if (std::optional<std::string> IntrPath =
+            TC.getDefaultIntrinsicModuleDir()) {
+      SmallString<128> P(*IntrPath);
```
build as std::string to start if you're returning std::string?

Meinersbur replied:
There is no `llvm::sys::path::append` overload for `std::string`. I'd prefer to use it for consistent handling of directory separator, trailing slashes, and absolute paths, etc.

Also note that this is the pattern used throughout this file.

Meinersbur committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

Meinersbur committed:
Merge commit '8e132f78bfb070effd247e85b19e3c84b9e84bac^' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge commit '8e132f78bfb070effd247e85b19e3c84b9e84bac' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/flang_builtin-mods_3' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Restore hiding module files deep in the build directory

Leaving up to CMake still clashed building the same source using
multiple targets.

Meinersbur commented:
@petrhosek Please review

Meinersbur committed:
Merge commit '8b96c2104e74f90d6431022001cac2dccdbdeed0^' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge commit '8b96c2104e74f90d6431022001cac2dccdbdeed0' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge commit '3174c94eaf2066a3b565e59a4f9d8d30444d2cd7' into users/meinersbur/flang_builtin-mods_3

petrhosek commented on runtimes/cmake/config-Fortran.cmake:
```diff
@@ -78,7 +78,7 @@ set(RUNTIMES_ENABLE_FORTRAN OFF)
 #    add_dependencies(target ${RUNTIMES_FORTRAN_BUILD_DEPS})
 #
-# to not fail
+# to noy fail
```
Is this an intentional change?

petrhosek reviewed (approved):
This PR touches a lot of different parts of LLVM which makes it difficult to review in its entirety. I think it'd be better to further break up this PR into smaller ones that could be landed separately to avoid having the revert & reland the entire change; for example, I believe the Clang driver changes could land separately. However, this is ultimately a call for maintainers of Clang, Flang and OpenMP, the build system changes look good to me.

tarunprabhu commented on flang-rt/cmake/modules/AddFlangRT.cmake:
```diff
@@ -225,13 +242,27 @@ function (add_flangrt_library name)
+        # Always enable preprocessor regardless of file extension
+        "$<$<COMPILE_LANGUAGE:Fortran>:-cpp>"
+
+        # Missing type descriptors are expected for intrinsic modules
```
Will type descriptors always be absent in intrinsic modules? Is this something that might affect end-users as well? This is not related to this PR, but is it worth exposing this as a driver option instead of using `-mmlir`?

Meinersbur replied:
As this was already only exposes as an `-mmlir` option, I think it was considered too internal to Flang to be considered as a driver option. This is OK for intrinsics since those are compiler-specific as well (Flang handles first-class types internally). 

In this case the option is necessary because before this PR, `-fsyntax-only` is used to build those modules, skipping object code generation. That is, those descriptors did never landed in the flang-rt anway.

tarunprabhu commented on clang/lib/Driver/ToolChains/Flang.cpp:
```diff
@@ -1179,6 +1179,37 @@ void Flang::ConstructJob(Compilation &C, const JobAction &JA,
+  // To reduce friction when build systems have not yet been updated, we also
+  // add the host's builtin module to the search path (with lower priority), in
+  // case a module file has not been found for the offload targets itself.
+  // FIXME: This workaround may mix module files targeting different triples and
```
Do you know what needs to happen before this code can be removed? If so, perhaps that should be made clear here so anyone looking at this code in the future can decide if it is an appropriate time to remove this workaround.

Meinersbur replied:
Downstream CI infrastructure needs to adjusted (`-DLLVM_RUNTIME_TARGETS=default;amdgcn-amd-amdhsa;nvptx64-nvidia-cuda` whenever `-DLLVM_ENABLE_RUNTIMES=openmp` and `-DLLVM_ENABLE_PROJECTS=offload`).

If you think this is not appropriately explained, please suggest how it should be written.

tarunprabhu replied:
Sorry, I still don't understand. I think my confusion stems from the use of the term "build systems" here. It is not clear if you mean other parts of LLVM's own build system, or if it means that applications _using_ flang need to adjust something.

Meinersbur replied:
Please read this PR's summary:
<img width="1047" height="327" alt="image" src="https://github.com/user-attachments/assets/c8454a8a-cfc6-410f-bc13-80c72987c04c" />

Build system is LLVM's CMake build system. That should already have been obvious because `LLVM_RUNTIME_TARGETS/LLVM_ENABLE_RUNTIMES/LLVM_ENABLE_PROJECTS` are LLVM-build-system-specific options.

tarunprabhu commented on clang/include/clang/Options/Options.td:
```diff
@@ -6431,7 +6431,7 @@ def prebind : Flag<["-"], "prebind">;
 def preload : Flag<["-"], "preload">;
 def print_file_name_EQ : Joined<["-", "--"], "print-file-name=">,
```
Enabling this command-line option really should be a separate PR, I'm afraid. The test can be modeled off of `clang/test/Driver/print-file-name.c`.

Meinersbur replied:
#195790

tarunprabhu reviewed (commented):
I mostly looked at the driver-related code. I think enabling the `-print-file-names` option really should be a separate PR. Other than that the driver related code looks fine.

cenewcombe commented on openmp/module/CMakeLists.txt:
```diff
@@ -6,84 +6,30 @@
-  )
-endif ()
+install(FILES
+  "${CMAKE_CURRENT_BINARY_DIR}/../runtime/src/omp_lib.h"
+  DESTINATION ${LIBOMP_HEADERS_INSTALL_PATH}
```
This is placing the fortran header file in the clang resource directory. Could you instead use `RUNTIMES_INSTALL_RESOURCE_MOD_PATH`, so that flang can find it with the fortran module files?

A quick test to demonstrate the problem/solution:
```
program main
  include "omp_lib.h"

   print *, 'PASS: openmp_version parameter ', openmp_version
end program main
```

Meinersbur replied:
`LIBOMP_HEADERS_INSTALL_PATH` is not the resource dir (it is `include` by default), and it the same path as before this PR, nor is `RUNTIMES_INSTALL_RESOURCE_MOD_PATH` in the default search path for includes. I do not want to overload this PR with additional unrelated changes.

Meinersbur commented:
> I think it'd be better to further break up this PR into smaller ones that could be landed separately to avoid having the revert & reland the entire change; for example, I believe the Clang driver changes could land separately. 

The driver change implements the new location of the module files. If landing only the driver change or only the module output location, the module files will not be found. I can try adding both locations to the search path, but I think it will cause more problems than it solves (`.mod` files contain the hash of the `.mod` files it depends on, and cause an error if the first match has a different hash)

tarunprabhu commented on clang/lib/Driver/ToolChains/Flang.cpp:
```diff
@@ -1179,6 +1179,37 @@ void Flang::ConstructJob(Compilation &C, const JobAction &JA,
+  auto &&HostTCs =
+      C.getOffloadToolChains<clang::driver::OffloadAction ::OFK_Host>();
+  for (auto [OKind, HostTC] : llvm::make_range(HostTCs.first, HostTCs.second)) {
+    if (HostTC == &TC)
+      continue;
+
+    if (std::optional<std::string> IntrModPath =
+            HostTC->getDefaultIntrinsicModuleDir()) {
+      CmdArgs.push_back("-fintrinsic-modules-path");
+      CmdArgs.push_back(Args.MakeArgString(*IntrModPath));
+    }
+  }
+
```
Sorry, somehow this didn't register in my earlier pass, but perhaps the following is clearer, since the body of the conditional is relatively short?

```suggestion
    if (HostTC != &TC) {
      if (std::optional<std::string> IntrModPath =
              HostTC->getDefaultIntrinsicModuleDir()) {
        CmdArgs.push_back("-fintrinsic-modules-path");
        CmdArgs.push_back(Args.MakeArgString(*IntrModPath));
      }
    }
```

tarunprabhu reviewed (commented):
Thanks for the other PR that adds `-print-file-name` support to Flang. Apologies, but I am still a little confused by the `FIXME` comment.

Meinersbur committed:
Merge branch 'main' (early part) into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
clang-format

Meinersbur committed:
Only assume native modules

Meinersbur committed:
Incorporate bbc change

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Fix path

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Add comment about module-only sources & rename RUNTIMES_ENABLE_FLANG_MODULES

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/flang_builtin-mods_3' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Fix typo

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Fix REAL16 support propagation

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/flang_builtin-mods_3' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Enable modules only with Fortran enabled

Meinersbur commented:
@tarunprabhu @vzakhari @DavidTruby @kkwli @DanielCChen All split-off PRs have landed and rebased to all changes in trunk, the latest being #197687 (see updated summary). Could you look at it again re-approve if everything is OK?
(This does not mean I would immediately land, will do an announcement on Discourse first)

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

jhuber6 commented on flang-rt/lib/runtime/CMakeLists.txt:
```diff
@@ -12,6 +12,31 @@ find_package(Backtrace)
+  ieee_exceptions.f90
+  ieee_features.f90
+  iso_c_binding.f90
+  iso_fortran_env.f90
```
This file is listed twice?

Meinersbur replied:
fixed

jhuber6 commented on flang-rt/lib/runtime/CMakeLists.txt:
```diff
@@ -220,6 +264,55 @@ if (TARGET FortranFloat128MathILib)
 
+
+# Build module files if requested.
+# The object files written by Flang for theses are unused. In the future parts
```
```suggestion
# The object files written by Flang for these are unused. In the future parts
```

jhuber6 commented on runtimes/cmake/config-Fortran.cmake:
```diff
@@ -261,7 +258,15 @@ function (flang_module_target tgtname)
+    # Modules of different targets must not overwrite each other.
+    # Ideally, we would use $<TARGET_PROPERTY:${tgtname},BINARY_DIR> but
+    # Fortran_MODULE_DIRECTORY does not support generator expressions.
+    # Not defining Fortran_MODULE_DIRECTORY at all would but it into
```
```suggestion
    # Not defining Fortran_MODULE_DIRECTORY at all would put it into
```

jhuber6 commented on flang/test/lit.cfg.py:
```diff
@@ -140,18 +140,91 @@
+    config.available_features.add("openmp_runtime")
+
+    # Search path for omp_lib.h with LLVM_ENABLE_RUNTIMES=openmp
+    # FIXME: In a bootstrpping build, openmp should write this file into a shared directory
```
```suggestion
    # FIXME: In a bootstrapping build, openmp should write this file into a shared directory
```

jhuber6 commented on flang-rt/lib/runtime/__fortran_builtins.f90:
```diff
@@ -6,7 +6,7 @@
 !===------------------------------------------------------------------------===!
 
-#include '../include/flang/Runtime/magic-numbers.h'
+#include "../../../flang/include/flang/Runtime/magic-numbers.h"
```
Why double quotes change?

jhuber6 commented on flang-rt/lib/runtime/CMakeLists.txt:
```diff
@@ -220,6 +264,55 @@ if (TARGET FortranFloat128MathILib)
+  flang_module_target(flang_rt.mod PUBLIC)
+
+  add_custom_target(flang-rt-mod
+    COMMAND ${CMAKE_COMMAND} -E true
```
Why do we need this? This is just an aggregate target, right? Also, do we really have an ordering on these anymore now that we don't build any Fortran?

Meinersbur replied:
See comment at line 286 for these same

This is for building modules which are still Fortran code. I think CMake tries to be clever in that if you have a dependency on the `flang_rt.mod` OBJECT library, it only builds the the object file that are actually needed by the dependee. Since the object files are not needed, it does not even build the module files. An `add_custom_target` is opaque to CMake and therefore stops CMake from being too clever.

Meinersbur replied:
I extracted the add_custom_command into a new function so the comment applies to any occurance where it is called.

jhuber6 reviewed (commented):
Overall LGTM, some nits

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Apply suggestions by @jhuber6

Meinersbur committed:
Merge remote-tracking branch 'official/users/meinersbur/flang_builtin-mods_3' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Typo fix

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_3

Meinersbur committed:
Flang does not understand -fvisibility=hidden -fapply-global-visibility-to-externs
